### PR TITLE
Switch to ShardsBatchGatewayAllocator in batch enabled mode for shard…

### PR DESCRIPTION
**Pull request by SwethaGuptha**
### Description
Change in AllocationService to switch to ShardsBatchGatewayAllocator for serving API _cluster/allocation/explain when batch mode is enabled.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
#13990

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java [hunk 1]
```diff
@@ -724,9 +724,17 @@ private AllocateUnassignedDecision explainUnassignedShardAllocation(ShardRouting
...
     private ExistingShardsAllocator getAllocatorForShard(ShardRouting shardRouting, RoutingAllocation routingAllocation) {
         assert assertInitialized();
-        final String allocatorName = ExistingShardsAllocator.EXISTING_SHARDS_ALLOCATOR_SETTING.get(
-            routingAllocation.metadata().getIndexSafe(shardRouting.index()).getSettings()
-        );
+        String allocatorName;
+        Boolean batchModeEnabled = EXISTING_SHARDS_ALLOCATOR_BATCH_MODE.get(settings);
+        if (batchModeEnabled
+            && routingAllocation.nodes().getMinNodeVersion().onOrAfter(Version.V_2_14_0)
+            && existingShardsAllocators.size() == 2) {
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java:731]: can you please create a private method for these conditions (present at 2 places now in same file)

File: server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java [hunk 2]
```diff
@@ -724,9 +724,17 @@ private AllocateUnassignedDecision explainUnassignedShardAllocation(ShardRouting
...
+            allocatorName = ShardsBatchGatewayAllocator.ALLOCATOR_NAME;
+        } else {
+            allocatorName = ExistingShardsAllocator.EXISTING_SHARDS_ALLOCATOR_SETTING.get(
+                routingAllocation.metadata().getIndexSafe(shardRouting.index()).getSettings()
+            );
+        }
         final ExistingShardsAllocator existingShardsAllocator = existingShardsAllocators.get(allocatorName);
```
Comment (rajiv-kv) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java:738]: `existingShardsAllocator` is getting referenced for start (`applyStartedShards`),failed(`applyFailedShards`) and count (`getNumberOfInFlightFetches`). Should we fix them also to use only the BatchAllocator if batch mode is enabled.

File: server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java (hunk 1 above)
Comment (SwethaGuptha) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java]: Added.

File: server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java (hunk 2 above)
Comment (SwethaGuptha) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java:732]: Yes, agreed. Discussed this with @shwetathareja, for decided on creating an issue for now and make an incremental change later.

**Comment by SwethaGuptha** [comment]
Failures: 
```
[org.opensearch.indices.IndicesRequestCacheIT.testDynamicStalenessThresholdUpdate {p0={"opensearch.experimental.feature.pluggable.caching.enabled":"false"}}](https://build.ci.opensearch.org/job/gradle-check/40439/testReport/junit/org.opensearch.indices/IndicesRequestCacheIT/testDynamicStalenessThresholdUpdate__p0___opensearch_experimental_feature_pluggable_caching_enabled___false___/)
    [org.opensearch.indices.IndicesRequestCacheIT.classMethod](https://build.ci.opensearch.org/job/gradle-check/40439/testReport/junit/org.opensearch.indices/IndicesRequestCacheIT/classMethod/)
```

Flaky test:

https://github.com/opensearch-project/OpenSearch/issues/13949

**Review by shwetathareja — APPROVED** [review]

File: server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java (hunk 2 above)
Comment (SwethaGuptha) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java:732]: Issue link: https://github.com/opensearch-project/OpenSearch/issues/14159

File: server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java [hunk 3]
```diff
@@ -724,13 +721,24 @@ private AllocateUnassignedDecision explainUnassignedShardAllocation(ShardRouting
...
         return existingShardsAllocator != null ? existingShardsAllocator : new NotFoundAllocator(allocatorName);
     }
 
+    private boolean isBatchModeEnabled(RoutingAllocation routingAllocation) {
+        return EXISTING_SHARDS_ALLOCATOR_BATCH_MODE.get(settings)
+            && routingAllocation.nodes().getMinNodeVersion().onOrAfter(Version.V_2_14_0)
+            && existingShardsAllocators.size() == 2;
```
Comment (rajiv-kv) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java:739]: we can move the `existingShardsAllocators.size() == 2` ahead of `routingAllocation.nodes().`

**Review by rajiv-kv — APPROVED** [review]

File: server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java (hunk 3 above)
Comment (SwethaGuptha) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java:739]: Discussed with @rajiv-kv that its not a blocking comment.

**Comment by SwethaGuptha** [comment]
Latest gradle run failure:

```
[Test Result](https://build.ci.opensearch.org/job/gradle-check/40647/testReport/) (1 failure / ±0)

    [org.opensearch.indices.IndicesRequestCacheIT.testStaleKeysRemovalWithoutExplicitThreshold {p0={"opensearch.experimental.feature.pluggable.caching.enabled":"false"}}](https://build.ci.opensearch.org/job/gradle-check/40647/testReport/junit/org.opensearch.indices/IndicesRequestCacheIT/testStaleKeysRemovalWithoutExplicitThreshold__p0___opensearch_experimental_feature_pluggable_caching_enabled___false___/)
```

flaky test: https://github.com/opensearch-project/OpenSearch/issues/14173

**Comment by SwethaGuptha** [comment]
> ❌ Gradle check result for [385dbe6](https://github.com/opensearch-project/OpenSearch/commit/385dbe66f753ede7a78a30e783db6269f6ee93f9): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/40701/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

 https://github.com/opensearch-project/OpenSearch/issues/14173

**Pull request merged**